src/storage/lock_manager/deadlock.rs

```diff
@@ -23,83 +23,116 @@ use std::fmt::{self, Display, Formatter};
+    // txn_ts => (lock_ts => (Vec<lock_hash>, last_detect_time))
+    wait_for_map: HashMap<u64, HashMap<u64, (Vec<u64>, time::Instant)>>,
+
+    /// The ttl of every edge.
+    ttl: time::Duration,
+
+    /// The time of last `active_expire`.
+    last_active_expire: time::Instant,
+
+    now: time::Instant,
 }
 
 impl DetectTable {
-    /// Returns the deadlock key hash which causes deadlock.
+    /// Creates a auto-expiring detect table.
+    pub fn new(ttl: time::Duration) -> Self {
+        Self {
+            wait_for_map: HashMap::default(),
+            ttl,
+            last_active_expire: time::Instant::now_coarse(),
+            now: time::Instant::now_coarse(),
+        }
+    }
+
+    /// Returns the key hash which causes deadlock.
     pub fn detect(&mut self, txn_ts: u64, lock_ts: u64, lock_hash: u64) -> Option<u64> {
         let _timer = DETECTOR_HISTOGRAM_VEC.detect.start_coarse_timer();
         TASK_COUNTER_VEC.detect.inc();
 
-        // If txn_ts is waiting for lock_ts, it won't cause deadlock.
+        self.now = time::Instant::now_coarse();
+        self.active_expire();
+
+        // If `txn_ts` is waiting for `lock_ts`, it won't cause deadlock.
         if self.register_if_existed(txn_ts, lock_ts, lock_hash) {
             return None;
         }
 
-        // Memorize the searching path.
-        let mut visited: HashSet<u64> = HashSet::default();
-        if let Some(deadlock_key_hash) = self.do_detect(&mut visited, txn_ts, lock_ts) {
+        if let Some(deadlock_key_hash) = self.do_detect(txn_ts, lock_ts) {
             return Some(deadlock_key_hash);
         }
         self.register(txn_ts, lock_ts, lock_hash);
         None
     }
 
-    /// Checks if there is a path from `wait_for_ts` to `txn_ts`.
-    fn do_detect(&self, visited: &mut HashSet<u64>, txn_ts: u64, wait_for_ts: u64) -> Option<u64> {
-        visited.insert(wait_for_ts);
-        if let Some(wait_for) = self.wait_for_map.get(&wait_for_ts) {
-            for (lock_ts, lock_hashes) in wait_for {
-                if *lock_ts == txn_ts {
-                    return Some(lock_hashes[0]);
-                }
-                if visited.contains(lock_ts) {
-                    continue;
-                }
-                if let Some(deadlock_key_hash) = self.do_detect(visited, txn_ts, *lock_ts) {
-                    return Some(deadlock_key_hash);
+    /// Checks if there is a edge from `wait_for_ts` to `txn_ts`.
+    fn do_detect(&mut self, txn_ts: u64, wait_for_ts: u64) -> Option<u64> {
+        let now = self.now;
+        let ttl = self.ttl;
+
+        // Memorize the searched vertexes.
+        let mut visited: HashSet<u64> = HashSet::default();
+        let mut stack = vec![wait_for_ts];
+        while let Some(wait_for_ts) = stack.pop() {
+            if let Some(wait_for) = self.wait_for_map.get_mut(&wait_for_ts) {
+                // Remove expired edges.
+                wait_for.retain(|_, (_, time)| now.duration_since(*time) < ttl);
```

Comment:
I think we can return a bool for the function active_expire, If we have already done the clean in the `active_expire` we do not need to check the expired edges here.

Reply:
I think it doesn't make much difference because active_expire run once per hour, but do_detect run many times in a second.